src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/Tensor.cs

```diff
@@ -0,0 +1,366 @@
+#pragma warning restore CA1825
+        }
+        // ITensor
+
+        /// <summary>
+        ///
+        /// </summary>
+        public static Tensor<T> Empty => EmptyTensor.Value;
+
+        /// <summary>
+        ///
+        /// </summary>
+        public bool IsEmpty => _lengths.Length == 0;
+
+        /// <summary>
+        ///
+        /// </summary>
+        public bool IsPinned => _isPinned;
+
+        /// <summary>
+        ///
+        /// </summary>
+        public int Rank => _lengths.Length;
+
+        /// <summary>
+        /// The number of items in the span.
+        /// </summary>
+        public nint Length => _linearLength;
+
+        // REIVEW: Calling this Shape for now as Lengths didn't seem to be the most liked option based on our discussion. Can rename if we desire.
+        /// <summary>
+        /// Gets the length of each dimension in this <see cref="Tensor{T}"/>.
+        /// </summary>
+        public ReadOnlySpan<nint> Shape => _lengths;
+
+        /// <summary>
+        /// Gets the strides of each dimension in this <see cref="Tensor{T}"/>.
+        /// </summary>
+        public ReadOnlySpan<nint> Strides => _strides;
+
+        /// <summary>
+        /// Returns a reference to specified element of the Span.
+        /// </summary>
+        /// <param name="indices"></param>
+        /// <returns></returns>
+        /// <exception cref="IndexOutOfRangeException">
+        /// Thrown when index less than 0 or index greater than or equal to Length
+        /// </exception>
+        public ref T this[params ReadOnlySpan<nint> indices] => ref AsSpanND()[indices];
+
+        // REVIEW: WE WILL WANT THIS CHANGED FROM A BOOL TO SOME FILTER EXPRESSION.
+        public Tensor<T> this[Tensor<bool> filter]
+        {
+            get
+            {
+                if (filter.Shape.Length != Shape.Length)
+                    throw new ArgumentOutOfRangeException(nameof(filter), "Number of dimensions to slice does not equal the number of dimensions in the span");
+
+                for (var i = 0; i < filter.Shape.Length; i++)
+                {
+                    if (filter.Shape[i] != Shape[i])
+                        ThrowHelper.ThrowArgument_FilterTensorMustEqualTensorLength();
+                }
+
+                Span<T> srcSpan = MemoryMarshal.CreateSpan(ref _values[0], (int)_linearLength);
+                Span<bool> filterSpan = MemoryMarshal.CreateSpan(ref filter._values[0], (int)_linearLength);
+
+                nint linearLength = SpanNDHelpers.CountTrueElements(filter);
+
+                T[] values = _isPinned ? GC.AllocateArray<T>((int)linearLength, _isPinned) : (new T[linearLength]);
+                int index = 0;
+                for (int i = 0; i < filterSpan.Length; i++)
+                {
+                    if (filterSpan[i])
+                    {
+                        values[i] = srcSpan[index++];
+                    }
+                }
```

Comment:
This seems like the type of thing that could just be reusing `InvokeSpanSpanIntoSpan` where `ROSpan<bool>` is reinterpreted as `ROSpan<byte>` instead